packages/core/js-sdk/src/client.ts

```diff
@@ -0,0 +1,183 @@
+import qs from "qs"
+import { ClientFetch, Config, FetchArgs, FetchInput, Logger } from "./types"
+
+const isBrowser = () => typeof window !== "undefined"
+
+const toBase64 = (str: string) => {
+  if (typeof window !== "undefined") {
+    return window.btoa(str)
+  }
+
+  return Buffer.from(str).toString("base64")
+}
+
+const sanitizeHeaders = (headers: Headers) => {
+  return {
+    ...Object.fromEntries(headers.entries()),
+    Authorization: "<REDACTED>",
+  }
+}
+
+const normalizeRequest = (
+  init: FetchArgs | undefined,
+  headers: Headers
+): RequestInit | undefined => {
+  let body = init?.body
+  if (body && headers.get("content-type")?.includes("application/json")) {
+    body = JSON.stringify(body)
+  }
+
+  return {
+    ...init,
+    headers,
+    ...(body ? { body: body as RequestInit["body"] } : {}),
+  } as RequestInit
+}
+
+const normalizeResponse = async (resp: Response, reqHeaders: Headers) => {
+  if (resp.status >= 300) {
+    const error = new FetchError(resp.statusText, resp.status)
+    throw error
+  }
+
+  // If we both requested JSON, we try to parse. Otherwise, we return the raw response.
+  const isJsonRequest = reqHeaders.get("accept")?.includes("application/json")
+  return isJsonRequest ? await resp.json() : resp
+}
+
+export class FetchError extends Error {
+  status: number | undefined
+
+  constructor(message: string, status?: number) {
+    super(`Status ${status}: ${message}`)
```

Comment:
suggestion: I think it would be better to just have the clean error message here,  what status a request failed with is irrelevant to the end user, they just want to see "Missing field 'id'" when the error is shown in a toast or something else.

Reply:
@kasperkristensen ah yes that is true. I added a ticket for us to actually return a JSON response so we can eg. localize the errors, but I agree with you for the time being.